gui/src/app/tinystan/index.ts

```diff
@@ -151,17 +158,21 @@ export default class StanModel {
     throw new Error(err_msg);
   }
 
+  private encodeInits(inits: string | object | string[] | object[]): cstr {
+    if (Array.isArray(inits)) {
+      return this.encodeString(inits.map((i) => jsonify(i)).join(this.sep));
+    } else {
+      return this.encodeString(jsonify(inits));
+    }
+  }
```

Comment:
I'm missing enough context to be clear--does this do the right thing if `inits` is an object (that might contain entries that would otherwise be multiple JSON values)? What are the circumstances under which it might be?

Maybe not possible at this point but it might be nice to give some type hinting for what this could be.

Reply:
I'm not able to follow what you're asking for, could you elaborate?

Reply:
Sure. So there are two cases here:

- `inits` is an `Array` (presumably of strings) -- then we encode each string (array element) in json and join them with the separator
- `inits` is not an `Array`--then we just encode it as one unit

But the typing suggests it could also be an (otherwise undetermined) `object`. I'm asking what potential use cases there would be when an object is passed, and would this logic cover those. Can we say anything about the shapes of valid object inputs here?

As a concrete example, `let isObj = [1, 2, 3]; typeof(isObj)` is 'object'. But `Array.isArray(isObj)` is also true, so we'd be JSONifying the individual values as separate JSON objects--which I assume isn't really what we want?

Conversely, if `inits` is `[[1, 2, 3], [4, 5, 6]]`, I expect this to return `[1, 2, 3] %SEP% [4, 5, 6]`, is that the correct behavior?

Reply:
Ah I follow now.

There are two "output behaviors" here, which is what we branch on:

1. We output a string containing a single json object (every chain gets this same initial value)
2. We output a string containing several json objects separated by a special character (different chains get different initial values)

The json objects in question will look like, in the simplest case, `{theta: 0.2}`. This corresponds to initializing the parameter called theta at 0.2. The values can be more complicated if you have e.g. a matrix parameter

As such, `[[1,2,3], [4,5,6]]` is invalid from the jump. I think until now I didn't realize the `object` type meant something like `Object` in Java - I was assuming it meant `object` in the JSON sense of key-value pairs.

If I change the type to 

```ts
type StanObject /* open to name suggestions */ = { [k:string]: any }
encodeInits(inits: string |StanObject | string[] | StanObject[])
```
is that more correct?

Reply:
Ahh, I get it. So we're always going to be using the name of the parameter plus its initial value.

I would probably use `StanParameterValue` instead of `StanObject` to be more semantic about it.

I think what you're really looking for is a [Record<string, number>](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type) or list of same, so maybe:

```typescript
type StanParameterValue = Record<string, number>
encodeInits(inits: string | StanParameterValue | string[] | StanParameterValue[])
```
?

Reply:
Two wrinkles -

1. I had already made this mistake before, this new type is also what `data` really is (which currently uses `object`, due to my same confusion). So the name should be a bit more generic
2. Record sounds good (and equivalent?), but `number` is too narrow.

Consider the stan model

```stan
parameters {
  vector[3] sigma;
}
```

A valid JSON is
```js
{
  "sigma": [1, 2, 3]
}
```

This can get arbitrarily complicated with tuple types in Stan now

```stan
tuple(real, matrix[2, 3] x;
```

would have a JSON value
```json
"x":  { "1" : 3.14, "2": [[1,2], [3,4], [5,6]]}
```

Reply:
> this new type is also what `data` really is

:+1:

> `number` is too narrow

Ah, I was afraid of that. Well, so long as the keys are serializable I suppose! If there's really no logic we need to apply here, we can just treat it as a blob.